src/utils/XFDR.cpp

```diff
@@ -0,0 +1,1095 @@
+       *
+       * @brief xprophet  method for target hits counting as implemented in xProphet
+       * @param cum_histograms Cumulative score distributions
+       * @param target_counts Number of target hits
+       */
+    /*
+      void target_xprophet(std::map< String, Math::CumulativeHistogram<>  * > & cum_histograms,
+                        Math::Histogram<> & target_counts)
+      {
+        // Required cumulative histograms for target hits
+        Math::CumulativeHistogram<>  intralinks_histogram = *cum_histograms[TOPPXFDR::xlclass_intralinks];
+        Math::CumulativeHistogram<>  interlinks_histogram = *cum_histograms[TOPPXFDR::xlclass_interlinks];
+        Math::CumulativeHistogram<>  monolinks_histogram = *cum_histograms[TOPPXFDR::xlclass_monolinks];
+
+        for (double current_score = TOPPXFDR::fpnum_score_start +  (TOPPXFDR::fpnum_score_step/2) ;
+             current_score <= TOPPXFDR::fpnum_score_end - (TOPPXFDR::fpnum_score_step/2);
+             current_score += TOPPXFDR::fpnum_score_step)
+        {
+            UInt n_intralinks = intralinks_histogram.binValue(current_score);
+            UInt n_interlinks = interlinks_histogram.binValue(current_score);
+            UInt n_monolinks = monolinks_histogram.binValue(current_score);
+            target_counts.inc(current_score, n_intralinks + n_interlinks + n_monolinks);
+        }
+     }
+     */
+    
+
+    /** Target counting as performed by the xProphet software package
+      *
+      * @brief xprophet  method for target hits counting as implemented in xProphet
+      * @param cum_histograms Cumulative score distributions
+      */
+    void fdr_xprophet(std::map< String, Math::CumulativeHistogram<>  * > & cum_histograms,
+                      const String  & targetclass, const String & decoyclass, const String & fulldecoyclass,
+                      vector< double > & fdr, bool mono)
+    {
+      for (double current_score = this->min_score +  (TOPPXFDR::fpnum_score_step/2) ;
+           current_score <= this->max_score - (TOPPXFDR::fpnum_score_step/2);
+           current_score += TOPPXFDR::fpnum_score_step)
+      {
+        double estimated_n_decoys = cum_histograms[decoyclass]->binValue(current_score);
+        if ( ! mono)
+        {
+          estimated_n_decoys -= 2 * cum_histograms[fulldecoyclass]->binValue(current_score);
+        }
+        double n_targets = cum_histograms[targetclass]->binValue(current_score);
+        fdr.push_back(n_targets > 0 ? estimated_n_decoys / (n_targets) : 0);
+      }
+    }
+
+    /**
+     * @brief Calculates the qFDR values for the provided FDR values, assuming that the FDRs are sorted by score in the input vector
+     * @param fdr Vector with FDR values which should be used for qFDR calculation
+     * @param qfdr Result qFDR values
+     */
+    void calc_qfdr(const vector< double > & fdr, vector< double > & qfdr)
+    {
+      qfdr.resize(fdr.size());
+      for (int i = fdr.size() - 1; i > -1; --i)
```

Comment:
I would write i >= 0 instead of i > -1
but I guess this is just a matter of taste